File: ios/Runner/QuranTranslation/TranslationStore.swift
import Foundation

/// On-disk lifecycle for one downloaded Quran translation language pack.
final class TranslationStore {
  static let shared = TranslationStore()

  private let fm = FileManager.default
  private let installer = AssetAtomicInstaller()

  private var rootURL: URL {
    let base = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
    return base.appendingPathComponent("QuranTranslations", isDirectory: true)
  }

  func languageRoot(_ languageCode: String) -> URL {
    rootURL.appendingPathComponent(languageCode, isDirectory: true)
  }

  func activeURL(_ languageCode: String) -> URL {
    languageRoot(languageCode).appendingPathComponent("active", isDirectory: true)
  }

  func translationFileURL(_ languageCode: String) -> URL {
    activeURL(languageCode).appendingPathComponent("translation.json")
  }

  func isAvailable(_ languageCode: String) -> Bool {
    fm.fileExists(atPath: translationFileURL(languageCode).path)
  }

  /// SHA-256 verify + atomic staging → active (keeps prior version in
  /// `previous/` until success). All downloaded languages are retained on disk.
  func installFromDirectory(
    _ stagingDirectory: URL,
    languageCode: String,
    manifest: [String: Any]
  ) throws {
    let translationName = (manifest["translation"] as? String) ?? "translation.json"
    let stagingFile = stagingDirectory.appendingPathComponent(translationName)
    guard fm.fileExists(atPath: stagingFile.path) else {
      throw AssetDownloadError.invalidManifest("Missing \(translationName) in staging.")
    }
    if let shaMap = manifest["sha256"] as? [String: Any],
      let expected = shaMap["translation"] as? String
    {
      try AssetIntegrityVerifier.verify(fileURL: stagingFile, expectedHex: expected)
    }

    let active = activeURL(languageCode)
    let previous = languageRoot(languageCode).appendingPathComponent("previous", isDirectory: true)

    let tempActive = languageRoot(languageCode)
      .appendingPathComponent("staging-active-\(UUID().uuidString)", isDirectory: true)
    try fm.createDirectory(at: tempActive, withIntermediateDirectories: true)
    try fm.copyItem(at: stagingFile, to: tempActive.appendingPathComponent("translation.json"))
    if let manifestData = try? JSONSerialization.data(withJSONObject: manifest, options: [.prettyPrinted]) {
      try manifestData.write(to: tempActive.appendingPathComponent("translation_manifest.json"))
    }

    try installer.activate(stagingDir: tempActive, activeDir: active, previousDir: previous)
  }

  private static let registerSharedCatalog: Void = {
    if AIAssetManager.shared.catalogURL == nil {
      AIAssetManager.shared.catalogURL = TranslationAssetDistributionConfig.catalogURL
    }
  }()

  func ensureTranslation(
    languageCode: String,
    progress: ((Double) -> Void)? = nil
  ) throws {
    _ = TranslationStore.registerSharedCatalog
    guard let entry = try TranslationCatalogCache.shared.entry(forLanguage: languageCode) else {
      throw AssetDownloadError.invalidManifest(
        "No translation_pack in catalog for language \"\(languageCode)\"."
      )
    }
    let plugin = TranslationAssetSync(assetId: entry.packId, languageCode: languageCode)
    AIAssetManager.shared.register(plugin)
    try AIAssetManager.shared.ensureAsset(assetId: entry.packId, progress: progress)
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/qurantranslation/TranslationStore.kt
package com.rnr.deenfocus.qurantranslation

import android.content.Context
import com.rnr.deenfocus.assetdownload.AIAssetManager
import com.rnr.deenfocus.assetdownload.AssetAtomicInstaller
import com.rnr.deenfocus.assetdownload.AssetDownloadException
import com.rnr.deenfocus.assetdownload.AssetIntegrityVerifier
import java.io.File
import org.json.JSONObject

/** On-disk lifecycle for one downloaded Quran translation language pack. */
class TranslationStore private constructor(context: Context) {
    private val appContext = context.applicationContext
    private val installer = AssetAtomicInstaller()

    private val rootDir: File
        get() = File(appContext.filesDir, "QuranTranslations")

    private val catalogCache: TranslationCatalogCache
        get() = TranslationCatalogCache(rootDir)

    fun languageRoot(languageCode: String): File = File(rootDir, languageCode)

    fun activeDir(languageCode: String): File = File(languageRoot(languageCode), "active")

    fun translationFile(languageCode: String): File = File(activeDir(languageCode), "translation.json")

    fun isAvailable(languageCode: String): Boolean = translationFile(languageCode).exists()

    fun installFromDirectory(stagingDir: File, languageCode: String, manifest: JSONObject) {
        val translationName = manifest.optString("translation", "translation.json")
        val stagingFile = File(stagingDir, translationName)
        if (!stagingFile.exists()) {
            throw AssetDownloadException.InvalidManifest("Missing $translationName in staging.")
        }
        val shaMap = manifest.optJSONObject("sha256")
        val expected = shaMap?.optString("translation", null)
        AssetIntegrityVerifier.verify(stagingFile, expected)

        val active = activeDir(languageCode)
        val previous = File(languageRoot(languageCode), "previous")
        val tempActive = File(languageRoot(languageCode), "staging-active-${System.currentTimeMillis()}")
        tempActive.mkdirs()
        stagingFile.copyTo(File(tempActive, "translation.json"), overwrite = true)
        File(tempActive, "translation_manifest.json").writeText(manifest.toString(2))

        installer.activate(tempActive, active, previous)
        tempActive.deleteRecursively()
    }

    fun ensureTranslation(languageCode: String, progress: ((Double) -> Unit)? = null) {
        registerSharedCatalog()
        val entry = catalogCache.entryForLanguage(languageCode)
            ?: throw AssetDownloadException.InvalidManifest(
                "No translation_pack in catalog for language \"$languageCode\".",
            )
        val plugin = TranslationAssetSync(this, entry.packId, languageCode)
        AIAssetManager.shared.register(plugin)
        AIAssetManager.shared.ensureAsset(entry.packId, progress = progress)
    }

    companion object {
        @Volatile
        private var instance: TranslationStore? = null

        fun getInstance(context: Context): TranslationStore =
            instance ?: synchronized(this) {
                instance ?: TranslationStore(context.applicationContext).also { instance = it }
            }

        private fun registerSharedCatalog() {
            if (AIAssetManager.shared.catalogUrl == null) {
                AIAssetManager.shared.catalogUrl = TranslationAssetDistributionConfig.catalogUrl
            }
        }
    }
}
